File: assessment-trainer/python/vision/pipeline.py
"""End-to-end screenshot analysis: preprocessing -> layout -> objects -> JSON."""
from __future__ import annotations

import time

import numpy as np

from .detect import Box, DetectionError, detect_layout
from .objects import extract_objects
from .preprocess import preprocess


def _to_original(b: Box | list[int], scale: float) -> list[int]:
    x, y, w, h = b.as_list() if isinstance(b, Box) else b
    return [int(round(v / scale)) for v in (x, y, w, h)]


def analyze(data: bytes) -> dict:
    timings: dict[str, float] = {}
    t0 = time.perf_counter()
    try:
        pre = preprocess(data)
    except ValueError as e:
        return {"ok": False, "stage": "decode", "error": str(e)}
    timings["preprocess_ms"] = (time.perf_counter() - t0) * 1000
    scale = pre.scale
    base = {
        "imageSize": [int(pre.original_shape[1]), int(pre.original_shape[0])],
        "preprocessing": pre.steps,
    }
    t1 = time.perf_counter()
    try:
        layout = detect_layout(pre.binary, pre.edges)
    except DetectionError as e:
        return {
            **base,
            "ok": False,
            "stage": e.stage,
            "error": str(e),
            "region": _to_original(e.region, scale) if e.region else None,
            "candidateBoxes": [_to_original(b, scale) for b in e.boxes[:60]],
            "timings": timings,
        }
    timings["detect_ms"] = (time.perf_counter() - t1) * 1000
    lat = layout.matrix
    if layout.missing is None:
        n_missing = len(lat.missing)
        return {
            **base,
            "ok": False,
            "stage": "missing_cell",
            "error": (
                "No missing cell could be identified (every grid slot contains a cell)."
                if n_missing == 0
                else f"{n_missing} grid slots have no detected cell; expected exactly one missing cell."
            ),
            "region": _to_original(lat.bbox, scale),
            "candidateBoxes": [_to_original(b, scale) for b in lat.cells.values()],
            "timings": timings,
        }
    t2 = time.perf_counter()
    cells: list[dict | None] = []
    debug_cells = []
    qualities = []
    for r in range(lat.rows):
        for c in range(lat.cols):
            box = lat.cells.get((r, c))
            if box is None:
                cells.append(None)
                debug_cells.append({"row": r, "col": c, "box": _to_original(lat.slot_box(r, c), scale), "missing": True, "objects": []})
                continue
            objs, q = extract_objects(pre.gray, (box.x, box.y, box.w, box.h))
            qualities.append(q)
            cells.append({"objects": [o.to_json() for o in objs]})
            debug_cells.append({
                "row": r, "col": c, "box": _to_original(box, scale), "missing": False,
                "objects": [{**o.to_json(), "confidence": round(o.confidence, 2),
                             "bbox": _to_original([box.x + o.bbox[0], box.y + o.bbox[1], o.bbox[2], o.bbox[3]], scale)} for o in objs],
            })
    options = []
    debug_options = []
    for b in layout.options:
        objs, q = extract_objects(pre.gray, (b.x, b.y, b.w, b.h))
        qualities.append(q)
        options.append({"objects": [o.to_json() for o in objs]})
        debug_options.append({
            "box": _to_original(b, scale),
            "objects": [{**o.to_json(), "confidence": round(o.confidence, 2),
                         "bbox": _to_original([b.x + o.bbox[0], b.y + o.bbox[1], o.bbox[2], o.bbox[3]], scale)} for o in objs],
        })
    timings["objects_ms"] = (time.perf_counter() - t2) * 1000

    # Extraction quality: object classification confidence, grid regularity,
    # and sanity checks (every cell/option contains something).
    sizes = [b.w for b in lat.cells.values()]
    regularity = 1.0 - min(1.0, float(np.std(sizes) / max(np.mean(sizes), 1)) * 5)
    empty = sum(1 for c in cells if c is not None and not c["objects"]) + sum(1 for o in options if not o["objects"])
    obj_q = float(np.mean(qualities)) if qualities else 0.0
    quality = obj_q * (0.7 + 0.3 * regularity) * (0.6 if empty else 1.0) * (0.85 if layout.inferred_options else 1.0)
    timings["total_ms"] = (time.perf_counter() - t0) * 1000
    mr, mc = layout.missing
    return {
        **base,
        "ok": True,
        "matrix": {"rows": lat.rows, "columns": lat.cols},
        "missingCell": [mr, mc],
        "answerOptions": len(options),
        "problem": {"rows": lat.rows, "cols": lat.cols, "cells": cells, "options": options},
        "quality": round(quality, 3),
        "qualityFactors": {"objects": round(obj_q, 3), "gridRegularity": round(regularity, 3), "emptyCells": empty, "inferredOptions": layout.inferred_options},
        "debug": {"matrixBox": _to_original(lat.bbox, scale), "cells": debug_cells, "options": debug_options},
        "timings": {k: round(v, 1) for k, v in timings.items()},
    }

File: assessment-trainer/lib/vision/js/pipeline.ts
/**
 * In-browser screenshot analysis: same stages and JSON contract as the Python
 * pipeline (python/vision/pipeline.py), implemented on plain pixel arrays.
 */
import type { Cell } from "../../matrigma/types";
import type { VisionResult } from "../types";
import { adaptiveInk, normalize, resize, toGray, type Gray } from "./raster";
import { DetectionError, detectLayout, latticeBox, latticeMissing, slotBox, type Box } from "./detect";
import { extractObjects } from "./objects";

export interface RgbaImage {
  width: number;
  height: number;
  data: ArrayLike<number>;
}

export function analyzeRgba(img: RgbaImage): VisionResult {
  const t0 = Date.now();
  const steps: string[] = ["decode"];
  let g: Gray = toGray(img.data, img.width, img.height);
  steps.push("grayscale");
  let scale = 1;
  const maxD = Math.max(g.w, g.h);
  const minD = Math.min(g.w, g.h);
  if (maxD > 2000) scale = 2000 / maxD;
  else if (minD < 600) scale = Math.min(2, 600 / minD);
  if (scale !== 1) {
    g = resize(g, Math.round(g.w * scale), Math.round(g.h * scale));
    steps.push(`resize x${scale.toFixed(2)}`);
  } else steps.push("resize (none needed)");
  g = normalize(g);
  steps.push("contrast normalisation");
  const block = Math.max(15, Math.floor(Math.min(g.w, g.h) / 40) | 1);
  const ink = adaptiveInk(g, block, 12);
  steps.push("adaptive threshold");
  const orig = (b: Box | number[]) => {
    const [x, y, w, h] = Array.isArray(b) ? b : [b.x, b.y, b.w, b.h];
    return [x, y, w, h].map((v) => Math.round(v / scale));
  };
  const base = { imageSize: [img.width, img.height] as [number, number], preprocessing: steps };
  let layout;
  try {
    layout = detectLayout(ink, g);
  } catch (e) {
    if (e instanceof DetectionError) {
      return {
        ...base,
        ok: false,
        stage: e.stage,
        error: e.message,
        region: e.region ? orig(e.region) : null,
        candidateBoxes: e.boxes.slice(0, 60).map((b) => orig(b)),
        timings: { total_ms: Date.now() - t0 },
      };
    }
    throw e;
  }
  const lat = layout.matrix;
  if (!layout.missing) {
    const n = latticeMissing(lat).length;
    return {
      ...base,
      ok: false,
      stage: "missing_cell",
      error: n === 0 ? "No missing cell could be identified (every grid slot contains a cell)." : `${n} grid slots have no detected cell; expected exactly one missing cell.`,
      region: orig(latticeBox(lat)),
      candidateBoxes: [...lat.cells.values()].map((b) => orig(b)),
      timings: { total_ms: Date.now() - t0 },
    };
  }
  let textureCells = 0;
  let wireCells = 0;
  let nestedCells = 0;
  const cells: (Cell | null)[] = [];
  const debugCells = [];
  const qualities: number[] = [];
  for (let r = 0; r < lat.rows; r++)
    for (let c = 0; c < lat.cols; c++) {
      const box = lat.cells.get(`${r},${c}`);
      if (!box) {
        cells.push(null);
        debugCells.push({ row: r, col: c, box: orig(slotBox(lat, r, c)), missing: true, objects: [] });
        continue;
      }
      const { objects, quality, texture, wire, nested } = extractObjects(g, box);
      if (texture) textureCells++;
      if (wire) wireCells++;
      if (nested) nestedCells++;
      qualities.push(quality);
      cells.push({ objects: objects.map(({ shape, fill, size, rotation, x, y }) => ({ shape, fill, size, rotation, x, y })) } as Cell);
      debugCells.push({
        row: r,
        col: c,
        box: orig(box),
        missing: false,
        objects: objects.map((o) => ({ ...o, bbox: orig([box.x + o.bbox[0], box.y + o.bbox[1], o.bbox[2], o.bbox[3]]) })),
      });
    }
  const options: Cell[] = [];
  const debugOptions = [];
  for (const b of layout.options) {
    const { objects, quality, texture, wire, nested } = extractObjects(g, b);
    if (texture) textureCells++;
    if (wire) wireCells++;
    if (nested) nestedCells++;
    qualities.push(quality);
    options.push({ objects: objects.map(({ shape, fill, size, rotation, x, y }) => ({ shape, fill, size, rotation, x, y })) } as Cell);
    debugOptions.push({ box: orig(b), objects: objects.map((o) => ({ ...o, bbox: orig([b.x + o.bbox[0], b.y + o.bbox[1], o.bbox[2], o.bbox[3]]) })) });
  }
  if (textureCells >= 2 || wireCells >= 3 || nestedCells >= 3) {
    textureCells = Math.max(textureCells, wireCells, nestedCells);
    return {
      ...base,
      ok: false,
      stage: "objects",
      error: `${textureCells} cells contain line patterns or overlapping figures, which the rule solver cannot read from an image.`,
      region: orig(latticeBox(lat)),
      candidateBoxes: [...lat.cells.values()].map((b) => orig(b)),
      timings: { total_ms: Date.now() - t0 },
    };
  }
  const sizes = [...lat.cells.values()].map((b) => b.w);
  const mean = sizes.reduce((s, v) => s + v, 0) / sizes.length;
  const std = Math.sqrt(sizes.reduce((s, v) => s + (v - mean) ** 2, 0) / sizes.length);
  const regularity = 1 - Math.min(1, (std / Math.max(mean, 1)) * 5);
  const empty = cells.filter((c) => c && !c.objects.length).length + options.filter((o) => !o.objects.length).length;
  const objQ = qualities.length ? qualities.reduce((s, v) => s + v, 0) / qualities.length : 0;
  const quality = objQ * (0.7 + 0.3 * regularity) * (empty ? 0.6 : 1) * (layout.inferredOptions ? 0.85 : 1);
  return {
    ...base,
    ok: true,
    matrix: { rows: lat.rows, columns: lat.cols },
    missingCell: layout.missing,
    answerOptions: options.length,
    problem: { rows: lat.rows, cols: lat.cols, cells, options },
    quality: Math.round(quality * 1000) / 1000,
    qualityFactors: { objects: objQ, gridRegularity: regularity, emptyCells: empty, inferredOptions: layout.inferredOptions },
    debug: { matrixBox: orig(latticeBox(lat)), cells: debugCells, options: debugOptions },
    timings: { total_ms: Date.now() - t0 },
    provider: "browser",
  };
}
